flink-runtime/src/main/java/org/apache/flink/runtime/util/DefaultGroupCache.java

```diff
@@ -0,0 +1,163 @@
+ */
+
+package org.apache.flink.runtime.util;
+
+import org.apache.flink.annotation.VisibleForTesting;
+
+import org.apache.flink.shaded.guava31.com.google.common.base.Ticker;
+import org.apache.flink.shaded.guava31.com.google.common.cache.Cache;
+import org.apache.flink.shaded.guava31.com.google.common.cache.CacheBuilder;
+import org.apache.flink.shaded.guava31.com.google.common.cache.RemovalNotification;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/** Default implement of {@link GroupCache}. Entries will be expired after timeout. */
+@NotThreadSafe
+public class DefaultGroupCache<G, K, V> implements GroupCache<G, K, V> {
+    private final Cache<CacheKey<G, K>, V> cache;
+    private final Map<G, Set<CacheKey<G, K>>> cachedBlobKeysPerJob;
+
+    private DefaultGroupCache(Duration expireTimeout, int cacheSizeLimit, Ticker ticker) {
+        this.cachedBlobKeysPerJob = new HashMap<>();
+        this.cache =
+                CacheBuilder.newBuilder()
+                        .concurrencyLevel(1)
+                        .maximumSize(cacheSizeLimit)
+                        .expireAfterAccess(expireTimeout)
+                        .ticker(ticker)
+                        .removalListener(this::onCacheRemoval)
+                        .build();
+    }
+
+    @Override
+    public void clear() {
+        cachedBlobKeysPerJob.clear();
+        cache.cleanUp();
+    }
+
+    @Override
+    public V get(G group, K key) {
+        return cache.getIfPresent(new CacheKey<>(group, key));
+    }
+
+    @Override
+    public void put(G group, K key, V value) {
+        CacheKey<G, K> cacheKey = new CacheKey<>(group, key);
+        cache.put(cacheKey, value);
+        cachedBlobKeysPerJob.computeIfAbsent(group, ignore -> new HashSet<>()).add(cacheKey);
+    }
+
+    @Override
+    public void clearCacheForGroup(G group) {
+        Set<CacheKey<G, K>> removed = cachedBlobKeysPerJob.remove(group);
+        if (removed != null) {
+            cache.invalidateAll(removed);
+        }
+    }
+
+    /**
+     * Removal listener that remove the cache key of this group .
+     *
+     * @param removalNotification of removed element.
+     */
+    private void onCacheRemoval(RemovalNotification<CacheKey<G, K>, V> removalNotification) {
+        CacheKey<G, K> cacheKey = removalNotification.getKey();
+        V value = removalNotification.getValue();
+        if (cacheKey != null && value != null) {
+            cachedBlobKeysPerJob.computeIfPresent(
+                    cacheKey.getGroup(),
+                    (group, keys) -> {
+                        keys.remove(cacheKey);
+                        if (keys.isEmpty()) {
+                            return null;
+                        } else {
+                            return keys;
+                        }
+                    });
```

Comment:
Would be there a risk of memory leakage here?
For example, let's talk about the situation:

- There are too many groups
- Perform the following operations on each of these groups one by one
  - Add a set for one group and then remove set for the same one group, but the key has not been removed. Would there be many Entries in the form of `Entry-i<Group-i, set-i>`(set-i is empty or null) ?
  
In short, would `cachedBlobKeysPerJob` degenerate into a collection with too many elements?

Please correct me if needed for my limited read.

Reply:
I write a demo on My IDEA, it doesn't have the memory leak. When return null, the map will remove the key.


<img width="1044" alt="image" src="https://github.com/apache/flink/assets/38427477/457dd9bb-9ea3-474f-aab4-b0c43e0b5020">


<img width="679" alt="image" src="https://github.com/apache/flink/assets/38427477/040b3831-1919-4a6e-afa0-8742e4bf200b">